extensions/obviousAlexC/penPlus.js

```diff
@@ -0,0 +1,2621 @@
+    },
+
+    getTextureData: (texture, width, height) => {
+      //?Initilize the temp framebuffer and assign it
+      const readBuffer = gl.createFramebuffer();
+
+      lastFB = gl.getParameter(gl.FRAMEBUFFER_BINDING);
+
+      gl.bindFramebuffer(gl.FRAMEBUFFER, readBuffer);
+
+      gl.framebufferTexture2D(
+        gl.FRAMEBUFFER,
+        gl.COLOR_ATTACHMENT0,
+        gl.TEXTURE_2D,
+        texture,
+        0
+      );
+
+      //?make sure to unbind the framebuffer and delete it!
+      const removeBuffer = () => {
+        gl.deleteFramebuffer(readBuffer);
+      };
+
+      //?if sucessful read
+      if (
+        gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE
+      ) {
+        //?Make an array to write the pixels onto
+        let dataArray = new Uint8Array(width * height * 4);
+        gl.readPixels(
+          0,
+          0,
+          width,
+          height,
+          gl.RGBA,
+          gl.UNSIGNED_BYTE,
+          dataArray
+        );
+
+        //?Remove Buffer data and return data
+        removeBuffer();
+        return dataArray;
+      }
+
+      //?If not return undefined
+      removeBuffer();
+      return undefined;
+    },
+
+    getTextureAsURI: (texture, width, height) => {
+      //?Initilize the temp framebuffer and assign it
+      const readBuffer = gl.createFramebuffer();
+
+      lastFB = gl.getParameter(gl.FRAMEBUFFER_BINDING);
+
+      gl.bindFramebuffer(gl.FRAMEBUFFER, readBuffer);
+
+      gl.framebufferTexture2D(
+        gl.FRAMEBUFFER,
+        gl.COLOR_ATTACHMENT0,
+        gl.TEXTURE_2D,
+        texture,
+        0
+      );
+
+      //?make sure to unbind the framebuffer and delete it!
+      const removeBuffer = () => {
+        gl.deleteFramebuffer(readBuffer);
+      };
+
+      //?if sucessful read
+      if (
+        gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE
+      ) {
+        //?Make an array to write the pixels onto
+        let dataArray = new Uint8Array(width * height * 4);
+        gl.readPixels(
+          0,
+          0,
+          width,
+          height,
+          gl.RGBA,
+          gl.UNSIGNED_BYTE,
+          dataArray
+        );
+
+        //Make an invisible canvas
+        const dataURICanvas = document.createElement("canvas");
+        dataURICanvas.width = width;
+        dataURICanvas.height = height;
+        const dataURIContext = dataURICanvas.getContext("2d");
+
+        // Copy the pixels to a 2D canvas
+        const imageData = dataURIContext.createImageData(width, height);
+        imageData.data.set(dataArray);
+        dataURIContext.putImageData(imageData, 0, 0);
+
+        //?Remove Buffer data and return data
+        removeBuffer();
+        return dataURICanvas.toDataURL();
+      }
+
+      //?If not return undefined
+      removeBuffer();
+      return undefined;
+    },
+  };
+
+  //Learned I could do this for code orginization
+  //?Menus
+  {
+    //? Define the menus
+    extension.addMenu(
```

Comment:
Shouldn't blocks and menus be defined he normal way?

Reply:
Its a workflow thing. just makes production of the extension faster since I don't have to flip between a long json list and a bunch of functions at the bottom of a class.

Reply:
If garbo does want me to make it normal syntax I am working on a converter though.